index.js

```diff
@@ -137,11 +138,58 @@ class Installer {
           this.pkgCount++
           return this
         })
+        .catch(e => {
+          if (child.optional) {
+            this.pkgCount++
+            child.failed = true
+            return rimraf(childPath)
+          } else {
+            throw e
+          }
+        })
       })
       return child.pending
     }, { concurrency: 50 })
   }
 
+  // A cute little mark-and-sweep collector!
+  garbageCollect (tree) {
+    const liveDeps = new Set()
+    const installer = this
+    mark(tree)
+    return sweep(tree)
+
+    function mark (tree) {
+      for (let dep of tree.dependencies.values()) {
+        if (dep.seenByMark) { continue }
+        dep.seenByMark = true
+        if (!dep.optional || !dep.failed) {
+          liveDeps.add(dep)
+          mark(dep)
+        }
+      }
+    }
+
+    function sweep (tree) {
+      return BB.map(tree.dependencies.values(), dep => {
+        if (dep.seenBySweep) { return }
+        dep.seenBySweep = true
+        return sweep(dep).then(() => {
+          if (!liveDeps.has(dep) && !dep.purged) {
+            const depPath = path.join(
+              installer.prefix,
+              'node_modules',
+              dep.address.replace(/:/g, '/node_modules/')
+            )
+            installer.pkgCount--
+            dep.purged = true
```

Comment:
Rather than marking as purged, if there was a failed Set, you could remove it from the Set here, and not have to track purged at all, I think.

Reply:
We're doing a tree crawl through the whole thing anyway. This wouldn't really affect much.